File: kairospy/application/usecases/risk/domain/budget.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from decimal import Decimal
from enum import StrEnum


class RiskMetric(StrEnum):
    NOTIONAL = "notional"
    MARGIN = "margin"
    GROSS_EXPOSURE = "gross_exposure"
    NET_EXPOSURE = "net_exposure"
    TURNOVER = "turnover"
    ORDER_RATE = "order_rate"


class RiskDecision(StrEnum):
    ALLOWED = "allowed"
    REDUCED = "reduced"
    REJECTED = "rejected"


class RiskReservationStatus(StrEnum):
    RESERVED = "reserved"
    PARTIALLY_CONSUMED = "partially_consumed"
    CONSUMED = "consumed"
    RELEASED = "released"
    EXPIRED = "expired"


@dataclass(frozen=True, slots=True)
class BudgetRef:
    scope: str
    subject: str

    def __post_init__(self) -> None:
        if not self.scope.strip() or not self.subject.strip():
            raise ValueError("budget scope and subject cannot be empty")


@dataclass(frozen=True, slots=True)
class RiskUsage:
    metric: RiskMetric
    amount: Decimal
    budgets: tuple[BudgetRef, ...]

    def __post_init__(self) -> None:
        if self.amount < 0:
            raise ValueError("risk usage cannot be negative")
        if not self.budgets:
            raise ValueError("risk usage requires at least one budget")


@dataclass(frozen=True, slots=True)
class RiskBudget:
    budget_id: str
    ref: BudgetRef
    metric: RiskMetric
    limit: Decimal
    used: Decimal = Decimal("0")
    reserved: Decimal = Decimal("0")
    valid_from: datetime | None = None
    valid_until: datetime | None = None

    def __post_init__(self) -> None:
        if not self.budget_id.strip():
            raise ValueError("budget_id cannot be empty")
        if self.limit < 0 or self.used < 0 or self.reserved < 0:
            raise ValueError("risk budget amounts cannot be negative")
        if self.used + self.reserved > self.limit:
            raise ValueError("risk budget usage exceeds limit")
        if self.valid_from and self.valid_from.tzinfo is None:
            raise ValueError("valid_from must be timezone-aware")
        if self.valid_until and self.valid_until.tzinfo is None:
            raise ValueError("valid_until must be timezone-aware")
        if self.valid_from and self.valid_until and self.valid_until < self.valid_from:
            raise ValueError("valid_until cannot precede valid_from")

    @property
    def available(self) -> Decimal:
        return self.limit - self.used - self.reserved

    def active_at(self, at: datetime) -> bool:
        if at.tzinfo is None:
            raise ValueError("risk evaluation timestamp must be timezone-aware")
        return (self.valid_from is None or at >= self.valid_from) and (
            self.valid_until is None or at < self.valid_until
        )


@dataclass(frozen=True, slots=True)
class RiskAllocation:
    budget_id: str
    metric: RiskMetric
    amount: Decimal


@dataclass(frozen=True, slots=True)
class RiskReservation:
    reservation_id: str
    request_id: str
    allocations: tuple[RiskAllocation, ...]
    status: RiskReservationStatus
    created_at: datetime


__all__ = [
    "BudgetRef",
    "RiskAllocation",
    "RiskBudget",
    "RiskDecision",
    "RiskMetric",
    "RiskReservation",
    "RiskReservationStatus",
    "RiskUsage",
]

File: crates/kairos-risk/src/domain/budget.rs
use std::cmp::Ordering;

use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct Amount {
    pub mantissa: i64,
    pub scale: u8,
}

impl Amount {
    pub const ZERO: Self = Self {
        mantissa: 0,
        scale: 0,
    };

    pub fn new(mantissa: i64, scale: u8) -> Result<Self, String> {
        if mantissa < 0 {
            return Err("risk amounts cannot be negative".into());
        }
        Ok(Self { mantissa, scale })
    }

    pub fn checked_add(self, other: Self) -> Result<Self, String> {
        let scale = self.scale.max(other.scale);
        let left = scale_value(self, scale)?;
        let right = scale_value(other, scale)?;
        let value = left
            .checked_add(right)
            .ok_or_else(|| "risk amount overflow".to_string())?;
        i64::try_from(value)
            .map(|mantissa| Self { mantissa, scale })
            .map_err(|_| "risk amount overflow".into())
    }

    pub fn checked_sub(self, other: Self) -> Result<Self, String> {
        let scale = self.scale.max(other.scale);
        let left = scale_value(self, scale)?;
        let right = scale_value(other, scale)?;
        let value = left
            .checked_sub(right)
            .ok_or_else(|| "risk amount underflow".to_string())?;
        if value < 0 {
            return Err("risk amount cannot become negative".into());
        }
        i64::try_from(value)
            .map(|mantissa| Self { mantissa, scale })
            .map_err(|_| "risk amount overflow".into())
    }

    pub fn cmp_value(self, other: Self) -> Ordering {
        let scale = self.scale.max(other.scale);
        scale_value(self, scale)
            .unwrap()
            .cmp(&scale_value(other, scale).unwrap())
    }
}

fn scale_value(value: Amount, scale: u8) -> Result<i128, String> {
    let factor = 10_i128
        .checked_pow((scale - value.scale) as u32)
        .ok_or_else(|| "risk scale overflow".to_string())?;
    (value.mantissa as i128)
        .checked_mul(factor)
        .ok_or_else(|| "risk amount overflow".into())
}

#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize, Deserialize)]
pub struct BudgetRef {
    pub scope: String,
    pub subject: String,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Metric {
    Notional,
    Margin,
    GrossExposure,
    NetExposure,
    Turnover,
    OrderRate,
}

impl Metric {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Notional => "notional",
            Self::Margin => "margin",
            Self::GrossExposure => "gross_exposure",
            Self::NetExposure => "net_exposure",
            Self::Turnover => "turnover",
            Self::OrderRate => "order_rate",
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct Budget {
    pub budget_id: String,
    pub owner_id: String,
    pub reference: BudgetRef,
    pub metric: Metric,
    pub limit: Amount,
    pub used: Amount,
    pub reserved: Amount,
    pub valid_from_unix_nanos: Option<u64>,
    pub valid_until_unix_nanos: Option<u64>,
}

impl Budget {
    pub fn validate(&self) -> Result<(), String> {
        if self.budget_id.trim().is_empty() || self.owner_id.trim().is_empty() {
            return Err("budget_id and owner_id are required".into());
        }
        if self.reference.scope.trim().is_empty() || self.reference.subject.trim().is_empty() {
            return Err("budget scope and subject are required".into());
        }
        if self
            .valid_from_unix_nanos
            .zip(self.valid_until_unix_nanos)
            .is_some_and(|(a, b)| b < a)
        {
            return Err("budget validity interval is inverted".into());
        }
        if self.used.checked_add(self.reserved)?.cmp_value(self.limit) == Ordering::Greater {
            return Err("budget usage exceeds limit".into());
        }
        Ok(())
    }
    pub fn active_at(&self, at: u64) -> bool {
        self.valid_from_unix_nanos.is_none_or(|v| at >= v)
            && self.valid_until_unix_nanos.is_none_or(|v| at < v)
    }
    pub fn available(&self) -> Amount {
        self.limit
            .checked_sub(self.used.checked_add(self.reserved).unwrap())
            .unwrap()
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct Usage {
    pub metric: Metric,
    pub amount: Amount,
    pub budgets: Vec<BudgetRef>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ReservationStatus {
    Reserved,
    Consumed,
    Released,
    Expired,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ReservationAllocation {
    pub budget_id: String,
    pub metric: Metric,
    pub amount: Amount,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct Reservation {
    pub reservation_id: String,
    pub request_id: String,
    pub allocations: Vec<ReservationAllocation>,
    pub status: ReservationStatus,
    pub created_at_unix_nanos: u64,
    pub updated_at_unix_nanos: u64,
}
